File: llmio-node/src/services/before.ts
export type Before = {
  model: string;
  stream: boolean;
  toolCall: boolean;
  structuredOutput: boolean;
  image: boolean;
  raw: Uint8Array;
};

function isObject(v: unknown): v is Record<string, unknown> {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

function getArray(v: unknown): unknown[] {
  return Array.isArray(v) ? v : [];
}

export function beforeOpenAI(raw: Uint8Array): Before {
  const json = JSON.parse(new TextDecoder().decode(raw)) as Record<string, unknown>;
  const model = String(json["model"] ?? "");
  if (!model) throw new Error("model is empty");

  const stream = Boolean(json["stream"]);
  if (stream) {
    const so = isObject(json["stream_options"]) ? json["stream_options"] : {};
    json["stream_options"] = { ...(so as object), include_usage: true };
    raw = new TextEncoder().encode(JSON.stringify(json));
  }

  const tools = getArray(json["tools"]);
  const toolCall = tools.length > 0;
  const structuredOutput = json["response_format"] !== undefined;

  let image = false;
  for (const msg of getArray(json["messages"])) {
    if (!isObject(msg)) continue;
    if (String(msg["role"] ?? "") !== "user") continue;
    const content = msg["content"];
    for (const part of getArray(content)) {
      if (isObject(part) && String(part["type"] ?? "") === "image_url") {
        image = true;
        break;
      }
    }
    if (image) break;
  }

  return { model, stream, toolCall, structuredOutput, image, raw };
}

export function beforeOpenAIRes(raw: Uint8Array): Before {
  const json = JSON.parse(new TextDecoder().decode(raw)) as Record<string, unknown>;
  const model = String(json["model"] ?? "");
  if (!model) throw new Error("model is empty");

  const stream = Boolean(json["stream"]);
  const tools = getArray(json["tools"]);
  const toolCall = tools.length > 0;

  const text = isObject(json["text"]) ? json["text"] : undefined;
  const format = text && isObject((text as Record<string, unknown>)["format"]) ? (text as Record<string, unknown>)["format"] : undefined;
  const structuredOutput = Boolean(format && isObject(format) && String((format as Record<string, unknown>)["type"] ?? "") === "json_schema");

  let image = false;
  for (const input of getArray(json["input"])) {
    if (!isObject(input)) continue;
    if (String(input["role"] ?? "") !== "user") continue;
    for (const part of getArray(input["content"])) {
      if (isObject(part) && String(part["type"] ?? "") === "input_image") {
        image = true;
        break;
      }
    }
    if (image) break;
  }

  return { model, stream, toolCall, structuredOutput, image, raw };
}

export function beforeAnthropic(raw: Uint8Array): Before {
  const json = JSON.parse(new TextDecoder().decode(raw)) as Record<string, unknown>;
  const model = String(json["model"] ?? "");
  if (!model) throw new Error("model is empty");

  const stream = Boolean(json["stream"]);
  const tools = getArray(json["tools"]);
  const toolCall = tools.length > 0;

  let image = false;
  for (const msg of getArray(json["messages"])) {
    if (!isObject(msg)) continue;
    if (String(msg["role"] ?? "") !== "user") continue;
    for (const part of getArray(msg["content"])) {
      if (isObject(part) && String(part["type"] ?? "") === "image") {
        image = true;
        break;
      }
    }
    if (image) break;
  }

  return { model, stream, toolCall, structuredOutput: toolCall, image, raw };
}

export function beforeGemini(raw: Uint8Array, model: string, stream: boolean): Before {
  if (!model) throw new Error("model is empty");
  const json = JSON.parse(new TextDecoder().decode(raw)) as Record<string, unknown>;

  const tools = getArray(json["tools"]);
  let toolCall = tools.length > 0;
  if (!toolCall && (json["toolConfig"] !== undefined || json["tool_config"] !== undefined)) toolCall = true;

  let structuredOutput = false;
  const generationConfig = isObject(json["generationConfig"])
    ? (json["generationConfig"] as Record<string, unknown>)
    : isObject(json["generation_config"])
      ? (json["generation_config"] as Record<string, unknown>)
      : undefined;
  const config = isObject(json["config"]) ? (json["config"] as Record<string, unknown>) : undefined;

  const responseJsonSchema =
    generationConfig?.["responseJsonSchema"] ??
    generationConfig?.["response_json_schema"] ??
    config?.["responseJsonSchema"] ??
    config?.["response_json_schema"];
  if (responseJsonSchema !== undefined) structuredOutput = true;

  const responseMimeType =
    generationConfig?.["responseMimeType"] ??
    generationConfig?.["response_mime_type"] ??
    config?.["responseMimeType"] ??
    config?.["response_mime_type"];
  if (String(responseMimeType ?? "").toLowerCase() === "application/json") structuredOutput = true;

  let image = false;
  for (const content of getArray(json["contents"])) {
    if (!isObject(content)) continue;
    for (const part of getArray(content["parts"])) {
      if (!isObject(part)) continue;
      const inlineData = isObject(part["inlineData"])
        ? (part["inlineData"] as Record<string, unknown>)
        : isObject(part["inline_data"])
          ? (part["inline_data"] as Record<string, unknown>)
          : undefined;
      const fileData = isObject(part["fileData"])
        ? (part["fileData"] as Record<string, unknown>)
        : isObject(part["file_data"])
          ? (part["file_data"] as Record<string, unknown>)
          : undefined;

      const mimeType = String(inlineData?.["mimeType"] ?? inlineData?.["mime_type"] ?? fileData?.["mimeType"] ?? fileData?.["mime_type"] ?? "");
      if (mimeType.startsWith("image/")) {
        image = true;
        break;
      }
    }
    if (image) break;
  }

  // 兼容 Go 版：Gemini 模型来自 URL 路径，不写回 body
  return { model, stream, toolCall, structuredOutput, image, raw };
}

File: service/before.go
package service

import (
	"errors"
	"strings"

	"github.com/tidwall/gjson"
	"github.com/tidwall/sjson"
)

type Before struct {
	Model            string
	Stream           bool
	toolCall         bool
	structuredOutput bool
	image            bool
	raw              []byte
}

type Beforer func(data []byte) (*Before, error)

// NewBeforerGemini 返回一个与现有 pre-processor 风格一致的 Gemini 解析器。
// 注意：Gemini 模型来自 URL 路径，不在 body 中。
func NewBeforerGemini(model string, stream bool) Beforer {
	return func(data []byte) (*Before, error) {
		if model == "" {
			return nil, errors.New("model is empty")
		}

		var toolCall bool
		if tools := gjson.GetBytes(data, "tools"); tools.Exists() && len(tools.Array()) != 0 {
			toolCall = true
		}
		if gjson.GetBytes(data, "toolConfig").Exists() || gjson.GetBytes(data, "tool_config").Exists() {
			toolCall = true
		}
		if !toolCall {
			gjson.GetBytes(data, "contents").ForEach(func(_, content gjson.Result) bool {
				if toolCall {
					return false
				}
				content.Get("parts").ForEach(func(_, part gjson.Result) bool {
					if part.Get("functionCall").Exists() || part.Get("function_call").Exists() ||
						part.Get("functionResponse").Exists() || part.Get("function_response").Exists() {
						toolCall = true
						return false
					}
					return true
				})
				return true
			})
		}

		var structuredOutput bool
		if gjson.GetBytes(data, "generationConfig.responseJsonSchema").Exists() ||
			gjson.GetBytes(data, "generationConfig.response_json_schema").Exists() ||
			gjson.GetBytes(data, "generation_config.responseJsonSchema").Exists() ||
			gjson.GetBytes(data, "generation_config.response_json_schema").Exists() ||
			gjson.GetBytes(data, "config.responseJsonSchema").Exists() ||
			gjson.GetBytes(data, "config.response_json_schema").Exists() {
			structuredOutput = true
		}
		if strings.EqualFold(gjson.GetBytes(data, "generationConfig.responseMimeType").String(), "application/json") ||
			strings.EqualFold(gjson.GetBytes(data, "generationConfig.response_mime_type").String(), "application/json") ||
			strings.EqualFold(gjson.GetBytes(data, "generation_config.responseMimeType").String(), "application/json") ||
			strings.EqualFold(gjson.GetBytes(data, "generation_config.response_mime_type").String(), "application/json") ||
			strings.EqualFold(gjson.GetBytes(data, "config.responseMimeType").String(), "application/json") ||
			strings.EqualFold(gjson.GetBytes(data, "config.response_mime_type").String(), "application/json") {
			structuredOutput = true
		}

		var image bool
		gjson.GetBytes(data, "contents").ForEach(func(_, content gjson.Result) bool {
			if image {
				return false
			}
			content.Get("parts").ForEach(func(_, part gjson.Result) bool {
				if image {
					return false
				}

				// 支持 camelCase 与 snake_case 两种字段命名
				inlineData := part.Get("inlineData")
				if !inlineData.Exists() {
					inlineData = part.Get("inline_data")
				}
				if inlineData.Exists() {
					mimeType := inlineData.Get("mimeType").String()
					if mimeType == "" {
						mimeType = inlineData.Get("mime_type").String()
					}
					if strings.HasPrefix(mimeType, "image/") {
						image = true
						return false
					}
				}

				fileData := part.Get("fileData")
				if !fileData.Exists() {
					fileData = part.Get("file_data")
				}
				if fileData.Exists() {
					mimeType := fileData.Get("mimeType").String()
					if mimeType == "" {
						mimeType = fileData.Get("mime_type").String()
					}
					if strings.HasPrefix(mimeType, "image/") {
						image = true
						return false
					}
				}
				return true
			})
			return true
		})

		return &Before{
			Model:            model,
			Stream:           stream,
			toolCall:         toolCall,
			structuredOutput: structuredOutput,
			image:            image,
			raw:              data,
		}, nil
	}
}

func BeforerOpenAI(data []byte) (*Before, error) {
	model := gjson.GetBytes(data, "model").String()
	if model == "" {
		return nil, errors.New("model is empty")
	}
	stream := gjson.GetBytes(data, "stream").Bool()
	if stream {
		// 为processTee记录usage添加选项 PS:很多客户端只会开启stream 而不会开启include_usage
		newData, err := sjson.SetBytes(data, "stream_options", struct {
			IncludeUsage bool `json:"include_usage"`
		}{IncludeUsage: true})
		if err != nil {
			return nil, err
		}
		data = newData
	}
	var toolCall bool
	tools := gjson.GetBytes(data, "tools")
	if tools.Exists() && len(tools.Array()) != 0 {
		toolCall = true
	}
	var structuredOutput bool
	if gjson.GetBytes(data, "response_format").Exists() {
		structuredOutput = true
	}
	var image bool
	gjson.GetBytes(data, "messages").ForEach(func(_, value gjson.Result) bool {
		if image {
			return false
		}
		if value.Get("role").String() == "user" {
			value.Get("content").ForEach(func(_, value gjson.Result) bool {
				if value.Get("type").String() == "image_url" {
					image = true
					return false
				}
				return true
			})
		}
		return true
	})
	return &Before{
		Model:            model,
		Stream:           stream,
		toolCall:         toolCall,
		structuredOutput: structuredOutput,
		image:            image,
		raw:              data,
	}, nil
}

func BeforerOpenAIRes(data []byte) (*Before, error) {
	model := gjson.GetBytes(data, "model").String()
	if model == "" {
		return nil, errors.New("model is empty")
	}
	stream := gjson.GetBytes(data, "stream").Bool()
	var toolCall bool
	tools := gjson.GetBytes(data, "tools")
	if tools.Exists() && len(tools.Array()) != 0 {
		toolCall = true
	}
	var structuredOutput bool
	if gjson.GetBytes(data, "text.format.type").String() == "json_schema" {
		structuredOutput = true
	}
	var image bool
	gjson.GetBytes(data, "input").ForEach(func(_, value gjson.Result) bool {
		if image {
			return false
		}
		if value.Get("role").String() == "user" {
			value.Get("content").ForEach(func(_, value gjson.Result) bool {
				if value.Get("type").String() == "input_image" {
					image = true
					return false
				}
				return true
			})
		}
		return true
	})
	return &Before{
		Model:            model,
		Stream:           stream,
		toolCall:         toolCall,
		structuredOutput: structuredOutput,
		image:            image,
		raw:              data,
	}, nil
}

func BeforerAnthropic(data []byte) (*Before, error) {
	model := gjson.GetBytes(data, "model").String()
	if model == "" {
		return nil, errors.New("model is empty")
	}
	stream := gjson.GetBytes(data, "stream").Bool()
	var toolCall bool
	tools := gjson.GetBytes(data, "tools")
	if tools.Exists() && len(tools.Array()) != 0 {
		toolCall = true
	}
	var image bool
	gjson.GetBytes(data, "messages").ForEach(func(_, value gjson.Result) bool {
		if image {
			return false
		}
		if value.Get("role").String() == "user" {
			value.Get("content").ForEach(func(_, value gjson.Result) bool {
				if value.Get("type").String() == "image" {
					image = true
					return false
				}
				return true
			})
		}
		return true
	})
	return &Before{
		Model:            model,
		Stream:           stream,
		toolCall:         toolCall,
		structuredOutput: toolCall,
		image:            image,
		raw:              data,
	}, nil
}
